src/lib/Generalize.hs

```diff
@@ -0,0 +1,231 @@
+generalizeArgs reqTys allArgs =  liftGeneralizerM $ runSubstReaderT idSubst do
+  PairE (Abs reqTys' UnitE) (ListE allArgs') <- sinkM $ PairE (Abs reqTys UnitE) (ListE allArgs)
+  ListE <$> go reqTys' allArgs'
+  where
+    go ::Nest PiBinder i i' -> [Atom n] -> SubstReaderT AtomSubstVal GeneralizerM i n [Atom n]
+    go Empty [] = return []
+    go (Nest (PiBinder b ty _) bs) (arg:args) = do
+      ty' <- substM ty
+      arg' <- case ty' of
+        TyKind   -> liftSubstReaderT $ generalizeType arg
+        DictTy _ -> generalizeDict ty' arg
+        _ -> return arg
+      args'' <- extendSubst (b@>SubstVal arg') $ go bs args
+      return $ arg' : args''
+    go _ _ = error "zip error"
+{-# INLINE generalizeArgs #-}
+
+-- === generalizer monad plumbing ===
+
+data GeneralizationEmission n l = GeneralizationEmission (Binder n l) (Atom n)
+type GeneralizationEmissions = RNest GeneralizationEmission
+
+newtype GeneralizerM n a = GeneralizerM {
+  runGeneralizerM' :: DoubleInplaceT Env GeneralizationEmissions UnitB HardFailM n a }
+  deriving (Functor, Applicative, Monad, MonadFail, Fallible, ScopeReader, EnvReader, EnvExtender)
+
+liftGeneralizerM
+  :: EnvReader m
+  => (forall l. DExt n l => GeneralizerM l (e l))
+  -> m n (Generalized e n)
+liftGeneralizerM cont = do
+  env <- unsafeGetEnv
+  Distinct <- getDistinct
+  Abs emissions (DoubleInplaceTResult UnitB e) <- return $ runHardFail $
+    runDoubleInplaceT env $ runGeneralizerM' cont
+  let (bs, vals) = hoistGeneralizationVals (unRNest emissions)
+  return (Abs bs e, vals)
+  where
+    -- TODO: something not O(N^2)
+    hoistGeneralizationVals :: Nest GeneralizationEmission n l -> (Nest Binder n l, [Atom n])
+    hoistGeneralizationVals Empty = (Empty, [])
+    hoistGeneralizationVals (Nest (GeneralizationEmission b val) bs) = do
+      let (bs', vals) = hoistGeneralizationVals bs
+      case hoist b (ListE vals) of
+        HoistSuccess (ListE vals') -> (Nest b bs', val:vals')
+        HoistFailure _ -> error "should't happen"
+{-# INLINE liftGeneralizerM #-}
+
+-- XXX: the supplied type may be more general than the type of the atom!
+emitGeneralizationParameter :: Type n -> Atom n -> GeneralizerM n (AtomName n)
+emitGeneralizationParameter ty val = GeneralizerM do
+  Abs b v <- return $ newName noHint
+  let emission = Abs (RNest REmpty (GeneralizationEmission (b:>ty) val)) v
+  emitDoubleInplaceTHoisted emission >>= \case
+    Nothing -> error $ "expected atom to be hoistable " ++ pprint val
+    Just v' -> return v'
+
+-- === actual generalization traversal ===
+
+-- Given a type (an Atom of type `Type`), abstracts over all data components
+generalizeType :: Type n -> GeneralizerM n (Type n)
+generalizeType ty = traverseTyParams ty \paramRole paramReqTy param -> case paramRole of
+  TypeParam -> generalizeType param
+  DictParam -> generalizeDict paramReqTy param
+  DataParam -> do
+    paramTyGeneralized <- getType param >>= generalizeType
```

Comment:
Why do you call `getType param` when you already have `paramReqTy`?

Reply:
Great observation. I think that should work. (Initially I didn't even call `generalizeType`, which was an error.) I'll try it.